### ISSUE TITLE
Cache not used if some caching keys aren't provided

### ISSUE THREAD

**Issue by ssimono** [aws/aws-appsync-community#224]
Hello. I have an AppSync GraphQL API connected to a lambda resolver.

I enabled per-resolver caching on a resolver with 3 cachingKeys, something like this:

```
cachingConfig: {
  ttl: core.Duration.minutes(60),
  cachingKeys: [
    '$context.arguments.query.area',
    '$context.arguments.query.model',
    '$context.arguments.query.discountCode',
  ],
}
```

The idea is that it computes some pricing, and caches the result per combination of those 3 fields. But let's say `discountCode` is optional, then every request without a discount code would never use the cache.

I would expect some caching entry for `area=Cologne, model=XXL, discountCode=undefined` to be created. But from the behavior I observed the cache seems to only be used if all 3 parameters are provided. I notice it from very clear differences in latency.

I tried to use something like `$util.defaultIfNull($context.arguments.query.discountCode, "NULL")` as one of the key using the template resolver syntax, but it doesn't allow because all caching keys must start with one of the top-level maps.

Anything I am missing? Is it the expected behavior?

Using aws-cdk 1.149.0

**Comment by dontepsu** [comment]
I'm also struggling with the same thing. Is there anyway we can solve this?

**Comment by onlybakam** [comment]
Hello, this is the correct behavior as we do strict matching on the keys and cannot run a large ranging scan to retrieve "subsets" of cache keys.

**Comment by markvp** [comment]
@onlybakam - your response suggests that you would be looking to clear "any cache item that has key A with value X"
I would like to be able to both cache and explicitly evict the following:

getValues(someId: ID, otherId: ID)

this could get called as getValues() - should be cached with key {someId: undefined/null, otherId: undefined/null} and therefore could be evicted with {} or {someId: undefined/null, otherId: undefined/null}

or it could be called as getValues(someId: 1) - which would get cached with key {someId: 1, otherId: undefined/null} and therefore could be evicted with {someId: 1} or {someId: 1, otherId: undefined/null}

etc...

I am not looking for a "large ranging scan to retrieve subsets of cache keys".

I do not want to pass {} and have it evict ALL ENTRIES, or even pass {someId: 1} and have it evict every entry that contains someId: 1... If I passed {} i would want it to ONLY evict the entry for getValues being called with no inputs, and if I evicted {someId: 1} I would want it to ONLY evict the entry for getValues being called with someId: 1 and no otherId defined.

Is there any reason that this could not be possible? it seems like a far simpler / clearer / explicit implementation that your suggested large ranging scan...
